File: src/main/java/preponderous/viron/services/EnvironmentService.java
// Copyright (c) 2024 Preponderous Software
// MIT License

package preponderous.viron.services;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import preponderous.viron.config.ServiceConfig;
import preponderous.viron.models.Environment;

@Service
public class EnvironmentService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final ServiceConfig serviceConfig;
    private final String baseUrl;

    @Autowired
    public EnvironmentService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder;

        this.serviceConfig = serviceConfig;

        this.baseUrl = this.serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/environment";
    }

    public List<Environment> getAllEnvironments() {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment[]> response = restTemplate.exchange(baseUrl + "/get-all-environments", HttpMethod.GET, null, Environment[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting environments");
        }
        return Arrays.asList(response.getBody());
    }

    public Environment getEnvironmentById(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/get-environment-by-id/" + id, HttpMethod.GET, null, Environment.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting environment by id");
        }
        return response.getBody();
    }

    public Environment getEnvironmentOfEntity(int entityId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/get-environment-of-entity/" + entityId, HttpMethod.GET, null, Environment.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting environment of entity");
        }
        return response.getBody();
    }

    public Environment createEnvironment(String name, int numGrids, int gridSize) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/create-environment/" + name + "/" + numGrids + "/" + gridSize, HttpMethod.POST, null, Environment.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error creating environment");
        }
        return response.getBody();
    }

    public boolean deleteEnvironment(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Boolean> response = restTemplate.exchange(baseUrl + "/delete-environment/" + id, HttpMethod.DELETE, null, Boolean.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error deleting environment");
        }
        return response.getBody();
    }

    public boolean setEnvironmentName(int id, String name) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Boolean> response = restTemplate.exchange(baseUrl + "/set-environment-name/" + id + "/" + name, HttpMethod.PUT, null, Boolean.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error setting environment name");
        }
        return response.getBody();
    }

}

File: src/main/python/preponderous/viron/services/environmentService.py

# Copyright (c) 2024 Preponderous Software
# MIT License

import requests
from src.main.python.preponderous.viron.models.environment import Environment

class EnvironmentService:
    def __init__(self, base_url):
        self.base_url = f"{base_url}/api/v1/environments"

    def get_all_environments(self) -> list[Environment]:
        response = requests.get(f"{self.base_url}")
        response.raise_for_status()
        return [Environment(**env) for env in response.json()]

    def get_environment_by_id(self, environment_id: int) -> Environment:
        response = requests.get(f"{self.base_url}/{environment_id}")
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_by_name(self, name: str) -> Environment:
        response = requests.get(f"{self.base_url}/name/{name}")
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_of_entity(self, entity_id: int) -> Environment:
        response = requests.get(f"{self.base_url}/entity/{entity_id}")
        response.raise_for_status()
        return Environment(**response.json())

    def create_environment(self, name: str, num_grids: int, grid_size: int) -> Environment:
        response = requests.post(f"{self.base_url}/{name}/{num_grids}/{grid_size}")
        response.raise_for_status()
        return Environment(**response.json())

    def delete_environment(self, environment_id: int) -> bool:
        response = requests.delete(f"{self.base_url}/{environment_id}")
        response.raise_for_status()
        return response.status_code == 200

    def update_environment_name(self, environment_id: int, name: str) -> bool:
        response = requests.patch(f"{self.base_url}/{environment_id}/name/{name}")
        response.raise_for_status()
        return response.status_code == 200